sklearn/neighbors/_base.py

```diff
@@ -542,24 +543,28 @@ def _fit(self, X, y=None):
                 else:
                     self._fit_method = "brute"
 
-        if self._fit_method == "ball_tree":
-            self._tree = BallTree(
-                X,
-                self.leaf_size,
-                metric=self.effective_metric_,
-                **self.effective_metric_params_,
-            )
-        elif self._fit_method == "kd_tree":
-            self._tree = KDTree(
-                X,
-                self.leaf_size,
-                metric=self.effective_metric_,
-                **self.effective_metric_params_,
-            )
-        elif self._fit_method == "brute":
-            self._tree = None
-        else:
-            raise ValueError("algorithm = '%s' not recognized" % self.algorithm)
+        with config_context(assume_finite=True):
+            # In the following cases, we remove the validation of X done at
+            # the beginning of the BinaryTree's constructors as X already got
+            # validated when calling this method, NeighborsBase._fit.
```

Comment:
Checks running time (represented by the small spike at the beginning) is negligible here.
![Screenshot 2021-08-06 at 18-04-18 Screenshot](https://user-images.githubusercontent.com/13029839/128539669-60d9b78c-9a1f-4323-ba2d-4dd609edcb22.png)

<details>

<summary> Script </summary>

```python
# profile.py
from sklearn.datasets import make_regression
from sklearn.neighbors import NearestNeighbors


def main(args=None):
    X_train, _ = make_regression(n_samples=100_000, n_features=10)
    X_test, _ = make_regression(n_samples=100_000, n_features=10)

    nn = NearestNeighbors(algorithm='kd_tree').fit(X_train)
    nn.kneighbors(X_test, n_neighbors=2)


if __name__ == "__main__":
    main()

```

```sh
giltracer --state-detect profile.py
```

</details>